From 7fdd5f74c211e3b677a6b85a34d0cc6de6be3d04 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Mon, 28 Oct 2024 14:19:26 +0100
Subject: [PATCH] Revert "media: v4l2-subdev: Document and enforce .s_stream()
 requirements"

This reverts commit 35556d0e1c61a25c737b9fc791a98add25464a8e.

Revert the commit introducing the WARN_ON within call_s_stream function
since current dcmipp driver behavior current requires to call s_stream
several times.  Preference is to remove this WARN_ON while waiting for
moving to the proper enable_streams / disable_streams ops within the
dcmipp driver.

Change-Id: I425d6b41551a0fb58bb0524adba97c5b97472404
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/421289
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Alain VOLMAT <alain.volmat@st.com>
Reviewed-by: Alain VOLMAT <alain.volmat@st.com>
Reviewed-by: Hugues FRUCHET <hugues.fruchet@st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
---
 drivers/media/v4l2-core/v4l2-subdev.c | 13 +------------
 include/media/v4l2-subdev.h           |  4 +---
 2 files changed, 2 insertions(+), 15 deletions(-)

--- a/drivers/media/v4l2-core/v4l2-subdev.c
+++ b/drivers/media/v4l2-core/v4l2-subdev.c
@@ -361,14 +361,6 @@ static int call_s_stream(struct v4l2_sub
 {
 	int ret;
 
-	/*
-	 * The .s_stream() operation must never be called to start or stop an
-	 * already started or stopped subdev. Catch offenders but don't return
-	 * an error yet to avoid regressions.
-	 */
-	if (WARN_ON(sd->s_stream_enabled == !!enable))
-		return 0;
-
 #if IS_REACHABLE(CONFIG_LEDS_CLASS)
 	if (!IS_ERR_OR_NULL(sd->privacy_led)) {
 		if (enable)
@@ -382,12 +374,9 @@ static int call_s_stream(struct v4l2_sub
 
 	if (!enable && ret < 0) {
 		dev_warn(sd->dev, "disabling streaming failed (%d)\n", ret);
-		ret = 0;
+		return 0;
 	}
 
-	if (!ret) {
-		sd->s_stream_enabled = enable;
-
 	return ret;
 }
 
--- a/include/media/v4l2-subdev.h
+++ b/include/media/v4l2-subdev.h
@@ -446,9 +446,7 @@ enum v4l2_subdev_pre_streamon_flags {
  * @s_stream: start (enabled == 1) or stop (enabled == 0) streaming on the
  *	sub-device. Failure on stop will remove any resources acquired in
  *	streaming start, while the error code is still returned by the driver.
- *	The caller shall track the subdev state, and shall not start or stop an
- *	already started or stopped subdev. Also see call_s_stream wrapper in
- *	v4l2-subdev.c.
+ *	Also see call_s_stream wrapper in v4l2-subdev.c.
  *
  * @g_pixelaspect: callback to return the pixelaspect ratio.
  *
